src/contracts/atlas/Atlas.sol

```diff
@@ -246,47 +247,67 @@ contract Atlas is Escrow, Factory {
+        uint256[] memory bidsAndIndices = new uint256[](solverOps.length);
+        uint256 zeroBidCount;
+        uint256 bidAmountFound;
+        uint256 bidsAndIndicesLastIndex = bidsAndIndices.length - 1; // computed once for efficiency
+
+        // First, get all bid amounts. Bids of zero are ignored by only storing non-zero bids in the array, from right
+        // to left. If there are any zero bids they will end up on the left as uint(0) values - in their sorted
+        // position. This reduces operations needed later when sorting the array in ascending order.
+        // Each non-zero bid amount is packed with its original solverOps array index, to fit into a uint256 value. The
+        // order of bidAmount and index is important - with bidAmount using the most significant bits, and considering
+        // we do not store zero bids in the array, the index values within the uint256 should not impact the sorting.
+
+        // |<------------------------- uint256 (256 bits) ------------------------->|
+        // |                                                                        |
+        // |<------------------ uint240 ----------------->|<-------- uint16 ------->|
+        // |                                              |                         |
+        // |                    bidAmount                 |          index          |
+        // |                                              |                         |
+        // |<------------------ 240 bits ---------------->|<------- 16 bits ------->|
+
+        for (uint256 i; i < solverOps.length; ++i) {
+            bidAmountFound = _getBidAmount(dConfig, userOp, solverOps[i], returnData, key);
+
+            if (bidAmountFound == 0 || bidAmountFound > type(uint240).max) {
+                // Zero bids are ignored: increment zeroBidCount offset
+                // Bids that would cause an overflow are also ignored
                 unchecked {
-                    ++j;
+                    ++zeroBidCount;
                 }
-                continue;
             } else {
-                bidAmounts[i] = bidPlaceholder;
-
-                for (uint256 k = i - j + 1; k > 0; k--) {
-                    if (bidPlaceholder > bidAmounts[sortedOps[k - 1]]) {
-                        sortedOps[k] = sortedOps[k - 1];
-                        sortedOps[k - 1] = i;
-                    } else {
-                        sortedOps[k] = i;
-                        break;
-                    }
-                }
+                // Non-zero bids are packed with their original solverOps index.
+                // The array is filled with non-zero bids from the right. This causes all zero bids to be on the left -
+                // in their sorted position, so fewer operations are needed in the sorting step below.
+                bidsAndIndices[bidsAndIndicesLastIndex - (i - zeroBidCount)] =
+                    uint256(bidAmountFound << BITS_FOR_INDEX | uint16(i));
```

Comment:
This will handle ties the opposite way compared to before. Because before a bid had to be strictly larger than the earlier bids to be shifted left:

```solidity
if (bidPlaceholder > bidAmounts[sortedOps[k - 1]]) {
    sortedOps[k] = sortedOps[k - 1];
    sortedOps[k - 1] = i;
} else {
    sortedOps[k] = i;
    break;
}
```

but now a later bid would have a higher `uint16(i)` value to break ties. Either way seems fine to me, so this is just a note